daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala

```diff
@@ -691,41 +702,38 @@ class DataProcessor private (
 
 class ParseResult(dp: DataProcessor, override val resultState: PState)
   extends DFDL.ParseResult
-  with WithDiagnosticsImpl
-  with ErrorHandler {
+  with WithDiagnosticsImpl {
 
   /**
-   * To be successful here, we need to capture xerces parse/validation
+   * To be successful here, we need to capture parse/validation
    * errors and add them to the Diagnostics list in the PState.
    *
-   * @param state the initial parse state.
+   * @param bytes the parsed Infoset
    */
   def validateResult(bytes: Array[Byte]): Unit = {
     Assert.usage(resultState.processorStatus eq Success)
-    val schemaURIStrings = resultState.infoset.asInstanceOf[InfosetElement].runtimeData.schemaURIStringsForFullValidation
-    try {
-      val bis = new java.io.ByteArrayInputStream(bytes)
-      Validator.validateXMLSources(schemaURIStrings, bis, this)
-    } catch {
-      //
-      // Some SAX Parse errors are thrown even if you specify an error handler to the
-      // validator.
-      //
-      // So we also need this catch
-      //
-      case e: SAXException =>
-        resultState.validationErrorNoContext(e)
+
+    val v = dp.validationMode match {
```

Comment:
If you hoist this val v into DataProcessor lazy val validator = .... then it will be computed once when first used, but reused by that data processor over and over if many parses are called which request validation.  You'd then pass that down to the actual validation. 

I'm not sure for Xerces if this will matter much, because we may have to construct a whole Xerces instance and initialize it, and read in the schema, for every thread. Maybe not. Depends on whether a Xerces validator once initialized can be called from multiple threads, and just returns a result. 

Still even if a whole separate xerces object is required per thread, then this could happen once per thread per DataProcessor, so it might be helpful if threads are long-living and used to process the same kind of data over and over - which would be typical of message-processing. 

But for Schematron, this could be a big win because the whole conversion of the schematron file into xslt, and construction of the transformer from the XSLT might be able to be done exactly once and reused. I am not sure about the transformer object being stateless or not. If not then that would have to be a ThreadLocal object, but at least the generation of the XSLT would be fully shared/once-only. 

v.validateXML will then be responsible for being thread-safe since parse can be called on many threads simultaneously, but that means v.validateXML will have to use ThreadLocal objects, as expected, for anything stateful.

Reply:
Re-reading what I just wrote.... not very clear. I think the point is that a validator's shared state, that can be shared across DataProcessor.parse calls requesting validation, is state of the DataProcessor object, and so should be constructed as a member of that object. 

Whether a given validator can in fact take advantage of this depends on the extent to which that validator "compiles" the schema/rules/etc into static data that is read-only, and then provides a validation API that is stateless.

Reply:
> I think the point is that a validator's shared state, that can be shared across DataProcessor.parse calls requesting validation, is state of the DataProcessor object, and so should be constructed as a member of that object.

That is what I took from the first post.  I havent been thinking about multiple runs with same processor as much as probably should have been, but I think your point is clear once that is considered.

Reply:
The catch to moving that up to a lazy in the DP is the `resultState.infoset.runtimeData.schemaURIStringsForFullValidation` access.

Reply:
The DataProcessor has the ssrd (SchemaSetRuntimeData) as a member.
The ssrd has an elementRuntimeData object which is the root element's ERD, and that has the schemaURIStringsForFullValidation member on it.  So you don't need the resultState object to get this.

Reply:
There are some other things going on there, are they needed

https://github.com/apache/incubator-daffodil/blob/master/daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala#L675-L677

Reply:
OK.  So the only issue is that `schemaURIStringsForFullValidation1` is private so it cant recurse if I copy that logic, otherwise it fits nicely in there.  Dont know if it runs.

Any thoughts on opening up `schemaURIStringsForFullValidation1`?  Ill commit it as is for review.